agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -33,68 +37,126 @@
    */ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     Collection<String> attendees = request.getAttendees();
-    ArrayList<TimeRange> times = new ArrayList<TimeRange>();
-     // minutes of day is + 1 to account for first and last minute of day
-    boolean[] minutes = new boolean[MINUTES_IN_DAY + 1];
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
 
-    for(int i = 0; i < minutes.length; i++) {
-      minutes[i] = true;
-    }
+    // minutes of day is + 1 to account for first and last minute of day
+    Availability[] minutes = new Availability[MINUTES_IN_DAY + 1];
+    // set all spots to default value of all available
+    Arrays.fill(minutes, Availability.ALL_AVAILABLE);
 
-    for(Event event: events) {
-      // if there's an overlap in attendees block off those times
-      if(attendeeOverlap(event.getAttendees(), attendees)) {
+    for (Event event: events) {
+      Availability status = Availability.ALL_AVAILABLE;
+      
+      if (! Collections.disjoint(event.getAttendees(), attendees)) {
+        // if there's overlap in mandatory attendees, time is unavailable
+        status = Availability.UNAVAILABLE;
+      } else if (! Collections.disjoint(event.getAttendees(), optionalAttendees)) {
+        // if there's an overlap in only the optional attendees, time is available for only the mandatory employees
+        status = Availability.MANDATORY_AVAILABLE;
+      }
+
+      if (status != Availability.ALL_AVAILABLE) {
         TimeRange range = event.getWhen();
 
-        for(int i = range.start(); i < range.end(); i++)
-          minutes[i] = false;
+        for (int i = range.start(); i < range.end(); i++)
+          // make sure unavailable times are not overwritten as mandatory available
+          if (minutes[i] != Availability.UNAVAILABLE)
```

Comment:
This check also seems like it would need to be updated if a new enum value is added.

Consider encapsulating this type of check within an enum method:

```
private enum Availability { 
  ALL_AVAILABLE,
  MANDATORY_AVAILABLE,
  UNAVAILABLE
  
  boolean isReplaceableBy(Availability availability) {
    switch (this) {
      case ALL_AVAILABLE:
        return true;
      case UNAVAILABLE:
        return false;
      case MANDATORY_AVAILABLE:
        return availability == ALL_AVAILABLE;
    }
  }
}
```

Then this check becomes:

```
  if (minutes[i].isReplaceableBy(status)) {
    minutes[i] = status;
  }
```

This would make the reason for this check clearer, and also be more resilient to changes to the enum.

Reply:
That makes sense. Thank you for clarifying this. This seems like a better practice. I will change it.

Reply:
Done